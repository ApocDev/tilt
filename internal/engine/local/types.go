package local

import (
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	"github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
)

type Cmd = v1alpha1.Cmd
type CmdList = v1alpha1.CmdList
type CmdStatus = v1alpha1.CmdStatus
type CmdSpec = v1alpha1.CmdSpec
type CmdStateWaiting = v1alpha1.CmdStateWaiting
type CmdStateTerminated = v1alpha1.CmdStateTerminated
type CmdStateRunning = v1alpha1.CmdStateRunning
type CmdServer = v1alpha1.CmdServer
type CmdServerSpec = v1alpha1.CmdServerSpec
type CmdServerStatus = v1alpha1.CmdServerStatus
type ObjectMeta = metav1.ObjectMeta
