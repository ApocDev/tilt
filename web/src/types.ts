export enum ResourceView {
  Log,
  Preview,
  Alerts,
}

export enum TriggerMode {
  TriggerModeAuto,
  TriggerModeManual,
}

export type Build = {
  Error: {} | string | null
  StartTime: string
  Log: string
  FinishTime: string
  Edits: Array<string> | null
  IsCrashRebuild: boolean
  Warnings: Array<string> | null
}

export type TiltBuild = {
  Version: string
  Date: string
  Dev: boolean
}

// what is the status of the resource in the cluster
export enum RuntimeStatus {
  Ok = "ok",
  Pending = "pending",
  Error = "error",
}

// What is the status of the resource with respect to Tilt
export enum ResourceStatus {
  BuildQueued, // in auto, if you have changed a file but an affected build hasn't started yet. In manual after you have clicked build, before it has started building
  Building,
  Error,
  Warning,
  Deploying,
  Deployed, // defer to RuntimeStatus
}

export type Resource = {
  Name: string
  CombinedLog: string
  BuildHistory: Array<any>
  CrashLog: string
  CurrentBuild: any
  DirectoriesWatched: Array<any>
  Endpoints: Array<string>
  PodID: string
  IsTiltfile: boolean
  LastDeployTime: string
  PathsWatched: Array<string>
  PendingBuildEdits: Array<string>
  PendingBuildReason: number
  PendingBuildSince: string
  ResourceInfo: {
    PodCreationTime: string
    PodLog: string
    PodName: string
    PodRestarts: number
    PodUpdateStartTime: string
    YAML: string
    PodStatus: string
    Endpoints: Array<string>
  }
  RuntimeStatus: string
  TriggerMode: TriggerMode
  HasPendingChanges: boolean
}
