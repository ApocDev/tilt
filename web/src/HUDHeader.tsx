import React, { PureComponent } from "react"
import styled from "styled-components"
import * as s from "./style-helpers"
import { SnapshotHighlight } from "./types"
import { ReactComponent as SnapshotSvg } from "./assets/svg/snapshot.svg"

type HUDHeaderProps = {
  podID?: string
  endpoints?: string[]
  podStatus?: string
  showSnapshotButton?: boolean
  handleOpenModal?: () => void
  highlight?: SnapshotHighlight | null
}

let Root = styled.div`
  display: flex;
  align-items: center;
  padding-top: ${s.SizeUnit(0.25)};
  padding-left: ${s.SizeUnit(0.5)};
  padding-right: ${s.SizeUnit(0.25)};
  background-color: ${s.Color.grayDarkest};
`
let ResourceInfo = styled.div`
  flex: 1;
  display: flex;
  padding-right: ${s.SizeUnit(0.5)};
`

let PodStatus = styled.span`
  font-family: ${s.Font.sansSerif};
  font-size: ${s.FontSize.small};
  flex: 1;
`

let PodId = styled.span``

let PortForward = styled.span``
let PortForwardLabel = styled.span`
  color: ${s.Color.grayLight};
  margin-left: ${s.SizeUnit(0.5)};
  border-left: 1px solid ${s.Color.gray};
  padding-left: ${s.SizeUnit(0.5)};
  margin-right: ${s.SizeUnit(0.25)};
`
let PortForwardLink = styled.a``


let SnapshotButton = styled.button`
  border: 1px solid transparent;
  font-family: ${s.Font.sansSerif};
  font-size: ${s.FontSize.smallest};
  background-color: transparent;
  color: ${s.Color.blue};
  display: flex;
  align-items: center;
  box-sizing: border-box;
  padding-left: ${s.SizeUnit(0.5)};
  padding-right: ${s.SizeUnit(0.5)};
  padding-top: ${s.SizeUnit(0.25)};
  padding-bottom: ${s.SizeUnit(0.25)};
  transition: border-color;
  transition-duration: ${s.AnimDuration.default};
  text-decoration: none;
  cursor: pointer;
  
  &:hover {
    background-color: ${s.Color.grayDark};
    border-color: ${s.Color.blue};
  }
  
  &.isHighlighted {
    border-color: ${s.Color.blue};
  }
`

let SnapshotButtonSvg = styled(SnapshotSvg)`
  margin-right: ${s.SizeUnit(0.25)};
`

class HUDHeader extends PureComponent<HUDHeaderProps> {
  renderSnapshotButton() {
    let highlight = this.props.highlight

    if (this.props.showSnapshotButton)
      return (
        <SnapshotButton
          onClick={this.props.handleOpenModal}
          className={`snapshotButton ${
            highlight ? "isHighlighted" : ""
          }`}
        >
          <SnapshotButtonSvg />
          <span>
            Create a <br />
            Snapshot
          </span>
        </SnapshotButton>
      )
  }

  render() {
    let podStatus = this.props.podStatus
    let podID = this.props.podID

    let endpoints = this.props.endpoints ?? []
    let endpointsEl = endpoints?.length > 0 && (
      <PortForward>
        <PortForwardLabel>
          Port-Forward{endpoints?.length > 1 ? "s" : ""}:
        </PortForwardLabel>
        {endpoints?.map(ep => (
          <PortForwardLink
            className="resourceInfo-value"
            href={ep}
            target="_blank"
            rel="noopener noreferrer"
            key={ep}
          >
            {ep}
          </PortForwardLink>
        ))}
      </PortForward>
    )

    return (
      <Root>
        <ResourceInfo>
          {podStatus &&
          <PodStatus>{podStatus}</PodStatus>
          }
          {podID &&
          <PodId>{podID}</PodId>
          }
          {endpointsEl}
        </ResourceInfo>
        {this.renderSnapshotButton()}
      </Root>
    )
  }
}

export default HUDHeader
