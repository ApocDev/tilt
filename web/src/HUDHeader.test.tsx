import React from "react"
import renderer from "react-test-renderer"
import { MemoryRouter } from "react-router"
import { ResourceView, SnapshotHighlight } from "./types"
import HUDHeader from "./HUDHeader"

const fakeHandleOpenModal = () => {}

it("shows snapshot url", () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <HUDHeader
          showSnapshotButton={true}
          handleOpenModal={fakeHandleOpenModal}
          highlight={null}
        />
      </MemoryRouter>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it("doesn't render snapshot button if it's a snapshot", () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <HUDHeader showSnapshotButton={false} />
      </MemoryRouter>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})
