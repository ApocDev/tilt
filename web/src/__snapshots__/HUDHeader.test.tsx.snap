// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doesn't render snapshot button if it's a snapshot 1`] = `
<div
  className="sc-bdVaJa lkbuwB"
>
  <div
    className="sc-bwzfXH fdpnBr"
  />
</div>
`;

exports[`shows snapshot url 1`] = `
<div
  className="sc-bdVaJa lkbuwB"
>
  <div
    className="sc-bwzfXH fdpnBr"
  />
  <button
    className="sc-gzVnrw ieJGxp snapshotButton "
    onClick={[Function]}
  >
    <svg
      className="sc-htoDjs jYtsma"
    >
      snapshot.svg
    </svg>
    <span>
      Create a 
      <br />
      Snapshot
    </span>
  </button>
</div>
`;
